swirl_proc_macro/src/background_job.rs

```diff
@@ -0,0 +1,209 @@
+    let fn_args = &job.args;
+    let struct_def = job.args.struct_def();
+    let struct_assign = job.args.struct_assign();
+    let arg_names = job.args.names();
+    let return_type = job.return_type;
+    let body = job.body;
+
+    let res = quote! {
+        #(#attrs)*
+        #vis #fn_token #name (#(#fn_args),*) -> #name :: Job {
+            #name :: Job {
+                #(#struct_assign),*
+            }
+        }
+
+        mod #name {
+            use super::*;
+
+            #[derive(serde::Serialize, serde::Deserialize)]
+            pub struct Job {
+                #(#struct_def),*
+            }
+
+            impl swirl::Job for Job {
+                type Environment = #env_type;
+                const JOB_TYPE: &'static str = stringify!(#name);
+
+                #fn_token perform(self, env: &Self::Environment) #return_type {
+                    let Job { #(#arg_names),* } = self;
+                    #(#body)*
+                }
+            }
+
+            swirl::register_job!(Job);
+        }
+    };
+    Ok(res)
+}
+
+struct BackgroundJob {
+    attrs: Vec<syn::Attribute>,
+    visibility: syn::Visibility,
+    fn_token: syn::Token![fn],
+    name: syn::Ident,
+    args: JobArgs,
+    return_type: syn::ReturnType,
+    body: Vec<syn::Stmt>,
+}
+
+impl BackgroundJob {
+    fn try_from(item: syn::ItemFn) -> Result<Self, Diagnostic> {
+        let syn::ItemFn {
+            attrs,
+            vis,
+            constness,
+            unsafety,
+            asyncness,
+            abi,
+            ident,
+            decl,
+            block,
+        } = item;
+
+        if let Some(constness) = constness {
+            return Err(constness
+                .span
+                .error("#[swirl::background_job] cannot be used on const functions"));
+        }
+
+        if let Some(unsafety) = unsafety {
+            return Err(unsafety
+                .span
+                .error("#[swirl::background_job] cannot be used on unsafe functions"));
+        }
+
+        if let Some(asyncness) = asyncness {
+            return Err(asyncness
+                .span
+                .error("#[swirl::background_job] cannot be used on async functions"));
+        }
+
+        if let Some(abi) = abi {
+            return Err(abi
+                .span()
+                .error("#[swirl::background_job] cannot be used on functions with an abi"));
+        }
+
+        if !decl.generics.params.is_empty() {
+            return Err(decl
+                .generics
+                .span()
+                .error("#[swirl::background_job] cannot be used on generic functions"));
+        }
+
+        if let Some(where_clause) = decl.generics.where_clause {
+            return Err(where_clause
+                .where_token
+                .span
+                .error("#[swirl::background_job] cannot be used on functions with a where clause"));
+        }
+
+        let fn_token = decl.fn_token;
```

Comment:
Since these are being immediately returned is it worth adding locals for them?

Reply:
`decl` is moved 2 lines down. We need to grab anything we need from it before that line